Modernize authn tasks for 1.5 API and auto-mTLS by default (#6546)
istio/istio.io: content/en/docs/tasks/security/authentication/authn-policy/index.md

### Setup

> Our examples use two namespaces `foo` and `bar`, with two services, `httpbin` and `sleep`, both running with an Envoy sidecar. We also use second

Reviewer: ```suggestion
Our examples use two namespaces `foo` and `bar`, with two services, `httpbin` and `sleep`, both running with an Envoy proxy. We also use second
```

Author: done

sleep.legacy to httpbin.legacy: 200
{{< /text >}}

> You should also verify that there is no peer authentication policy in the system, which you can do as follows:

Reviewer: ```suggestion
Verify there is no peer authentication policy in the system with the following command:
```

{{< text bash >}}
$ kubectl get peerauthentication --all-namespaces
No resources found.
{{< /text >}}

Last but not least, verify that there are no destination rules that apply on the example services. You can do this by checking the `host:` value of
 existing destination rules and make sure they do not match. For example:

{{< text bash >}}
$ kubectl get destinationrules.networking.istio.io --all-namespaces -o yaml | grep "host:"
{{< /text >}}

{{< tip >}}
Depending on the version of Istio, you may see destination rules for hosts other then those shown. However, there should be none with hosts in the `foo`,
`bar` and `legacy` namespace, nor is the match-all wildcard `*`
{{< /tip >}}

## Auto mTLS

> By default, Istio tracks the server workloads migrated to Istio sidecar, and configures client sidecar to send mutual TLS traffic to those workloads automatically, and send plain text traffic to workloads without sidecars.

Reviewer: ```suggestion
By default, Istio tracks the server workloads migrated to Istio proxies, and configures client proxies to send mutual TLS traffic to those workloads automatically, and to send plain text traffic to workloads without sidecars.
```

Author: done

> As a result, all traffic between workloads with sidecar will be in mTLS, without you do anything. To demonstrate that, let's examine the response from request to `httpbin/header`. When mTLS is in used, `X-Forwarded-Client-Cert` header will be injected by proxy sidecar to the upstream request to backend. Thus, exisent of that header is an evidence that mTLS is used. For example:

Reviewer: ```suggestion
Thus, all traffic between workloads with proxies uses mTLS, without you doing anything. For example, take the response from a request to `httpbin/header`. When using mTLS, the proxy injects the `X-Forwarded-Client-Cert` header to the upstream request to the backend. That header's presence is evidence that mTLS is used. For example:
```

Author: done

{{< text bash >}}
$ kubectl exec $(kubectl get pod -l app=sleep -n foo -o jsonpath={.items..metadata.name}) -c sleep -n foo -- curl http://httpbin.foo:8000/headers -s | grep X-Forwarded-Client-Cert
"X-Forwarded-Client-Cert": "By=spiffe://cluster.local/ns/foo/sa/httpbin;Hash=<redacted>"
{{< /text >}}

> On the contrary, if the server doesn't have sidecar, that header doesn't exist, implying request is in plaintext.

Reviewer: ```suggestion
When the server doesn't have sidecar, the `X-Forwarded-Client-Cert` header is not there, which implies requests are in plain text.
```

Author: done

sleep.legacy to httpbin.legacy: 200
{{< /text >}}

Verify there is no peer authentication policy in the system with the following command:

{{< text bash >}}
$ kubectl get peerauthentication --all-namespaces
No resources found.
{{< /text >}}

Last but not least, verify that there are no destination rules that apply on the example services. You can do this by checking the `host:` value of
 existing destination rules and make sure they do not match. For example:

{{< text bash >}}
$ kubectl get destinationrules.networking.istio.io --all-namespaces -o yaml | grep "host:"
{{< /text >}}

{{< tip >}}
Depending on the version of Istio, you may see destination rules for hosts other then those shown. However, there should be none with hosts in the `foo`,
`bar` and `legacy` namespace, nor is the match-all wildcard `*`
{{< /tip >}}

> ## Auto mTLS

Reviewer: ```suggestion
## Auto mutual TLS
```

Reviewer: I've checked and most spelling errors are caused because we don't abbreviate mutual TLS. 
Please replace all instances of "mTLS" with "mutual TLS"

Thus, all traffic between workloads with proxies uses mutual TLS, without you doing
anything. For example, take the response from a request to `httpbin/header`.
When using mutual TLS, the proxy injects the `X-Forwarded-Client-Cert` header to the
upstream request to the backend. That header's presence is evidence that mutual TLS is
used. For example:

{{< text bash >}}
$ kubectl exec $(kubectl get pod -l app=sleep -n foo -o jsonpath={.items..metadata.name}) -c sleep -n foo -- curl http://httpbin.foo:8000/headers -s | grep X-Forwarded-Client-Cert
"X-Forwarded-Client-Cert": "By=spiffe://cluster.local/ns/foo/sa/httpbin;Hash=<redacted>"
{{< /text >}}

When the server doesn't have sidecar, the `X-Forwarded-Client-Cert` header is not there, which implies requests are in plain text.

{{< text bash >}}
$ kubectl exec $(kubectl get pod -l app=sleep -n foo -o jsonpath={.items..metadata.name}) -c sleep -n foo -- curl http://httpbin.legacy:8000/headers -s | grep X-Forwarded-Client-Cert
{{< /text >}}

## Globally enabling Istio mutual TLS in STRICT mode

> As showing in the last section, while Istio automatically upgrade all traffic to mutual TLS between services with sidecar, it still allows services to receive plaintext traffic. If you want to prevent non-mutual TLS for the whole mesh, you can set a mesh-wide peer authentication policy to set mutual TLS mode to `STRICT`. Mesh-wide peer authentication policy must have empty `selector`, and defined in the *root namespace* like. For example:

Reviewer: ```suggestion
While Istio automatically upgrades all traffic between the proxies and the workloads to mutual TLS between, workloads can still receive plain text traffic. To prevent non-mutual TLS for the whole mesh, set a mesh-wide peer authentication policy to set mutual TLS mode to `STRICT`. The mesh-wide peer authentication policy shouldn't have a `selector` section, and it must apply to the **root namespace**, for example:
```

sleep.legacy to httpbin.legacy: 200
{{< /text >}}

Verify there is no peer authentication policy in the system with the following command:

{{< text bash >}}
$ kubectl get peerauthentication --all-namespaces
No resources found.
{{< /text >}}

Last but not least, verify that there are no destination rules that apply on the example services. You can do this by checking the `host:` value of
 existing destination rules and make sure they do not match. For example:

{{< text bash >}}
$ kubectl get destinationrules.networking.istio.io --all-namespaces -o yaml | grep "host:"
{{< /text >}}

{{< tip >}}
Depending on the version of Istio, you may see destination rules for hosts other then those shown. However, there should be none with hosts in the `foo`,
`bar` and `legacy` namespace, nor is the match-all wildcard `*`
{{< /tip >}}

## Auto mTLS

> By default, Istio tracks the server workloads migrated to Istio proxies, and configures client proxies to send mutual TLS traffic to those workloads automatically, and to send plain text traffic to workloads without sidecars.

Reviewer 2: we use `Istio proxies` in the first sentence and `without sidecars` in the latter one. The rest also use `proxies/sideacar`, suggest to stick to one of them consistently

Reviewer 2: (perfer) sidecar.

Reviewer 2: unless for a good reason, I suggest to use sidecar. reasons:

https://istio.io/docs/reference/glossary/ usage of sidecar is more common, proxy e.g. can be referred to egress gateway, https://istio.io/docs/reference/glossary/#tls-origination

doing search in istio.io website give me similar correlation

https://istio.io/search/?q=sidecar
https://istio.io/search/?q=proxy

Reviewer: We have tried to move away from using sidecar in favor of proxies. The reason is that while all sidecars have proxies, not all proxies are in sidecars. If I missed to replace an instance of "sidecar" with "proxy" please make the change.
As you point out, there are still a lot of mentions of sidecars that must be scrutinized to ensure they are accurate.
This is a longer conversation that I'm happy to continue post release.

Thus, all traffic between workloads with proxies uses mTLS, without you doing
anything. For example, take the response from a request to `httpbin/header`.
When using mTLS, the proxy injects the `X-Forwarded-Client-Cert` header to the
upstream request to the backend. That header's presence is evidence that mTLS is
used. For example:

{{< text bash >}}
$ kubectl exec $(kubectl get pod -l app=sleep -n foo -o jsonpath={.items..metadata.name}) -c sleep -n foo -- curl http://httpbin.foo:8000/headers -s | grep X-Forwarded-Client-Cert
"X-Forwarded-Client-Cert": "By=spiffe://cluster.local/ns/foo/sa/httpbin;Hash=<redacted>"
{{< /text >}}

When the server doesn't have sidecar, the `X-Forwarded-Client-Cert` header is not there, which implies requests are in plain text.

{{< text bash >}}
$ kubectl exec $(kubectl get pod -l app=sleep -n foo -o jsonpath={.items..metadata.name}) -c sleep -n foo -- curl http://httpbin.legacy:8000/headers -s | grep X-Forwarded-Client-Cert
{{< /text >}}

## Globally enabling Istio mutual TLS in STRICT mode

> As showing in the last section, while Istio automatically upgrade all traffic to mTLS between services with sidecar, it still allows services to receive plaintext traffic. If you want to prevent non-mTLS for the whole mesh, you can set a mesh-wide peer authentication policy to set mTLS mode to `STRICT`. Mesh-wide peer authentication policy must have empty `selector`, and defined in the *root namespace* like. For example:

Reviewer 2: /s/showing/showed

> As showing in the last section, while Istio automatically upgrade all traffic to mTLS between services with sidecar, it still allows services to receive plaintext traffic. If you want to prevent non-mTLS for the whole mesh, you can set a mesh-wide peer authentication policy to set mTLS mode to `STRICT`. Mesh-wide peer authentication policy must have empty `selector`, and defined in the *root namespace* like. For example:

Reviewer 2: upgrades

sleep.legacy to httpbin.bar: 000
command terminated with exit code 56
sleep.legacy to httpbin.legacy: 200
{{< /text >}}

You see requests still succeed, except for those from the client that doesn't have proxy, `sleep.legacy`, to the server with a proxy, `httpbin.foo` or `httpbin.bar`. This is expected because mTLS is now strictly required, but the workload without sidecar cannot comply.

### Cleanup part 1

Remove global authentication policy and destination rules added in the session:

{{< text bash >}}
$ kubectl delete peerauthentication -n istio-system default
{{< /text >}}

## Enable mutual TLS per namespace or workload

### Namespace-wide policy

To change mTLS for all workloads within a particular namespace, use a namespace-wide policy. The specification of the policy is the same as for a mesh-wide policy, but you specify the namespace it applies to under `metadata`. For example, the following peer authentication policy enables strict mTLS  for the `foo` namespace:

{{< text bash >}}
$ kubectl apply -f - <<EOF
apiVersion: "security.istio.io/v1beta1"
kind: "PeerAuthentication"
metadata:
  name: "default"
  namespace: "foo"
spec:
  mtls:
    mode: STRICT
EOF
{{< /text >}}

> As this policy is applied on services in namespace `foo` only, you should see only request from client-without-sidecar (`sleep.legacy`) to `httpbin.foo` start to fail.

Reviewer 2: s/services/workloads

Reviewer: Yes, workloads is better.

## Globally enabling Istio mutual TLS

To set a mesh-wide authentication policy that enables mutual TLS, submit peer authentication policy in the *root namespace* like below:

{{< text bash >}}
$ kubectl apply -f - <<EOF
apiVersion: "security.istio.io/v1beta1"
kind: "PeerAuthentication"
metadata:
  name: "default"
  namespace: "istio-system"
spec:
  mtls:
    mode: STRICT
EOF
{{< /text >}}

{{< tip >}}
The example above assumes the root namespace is `istio-system`. Please change it to the value you use with your Istio installation.
 {{< /tip >}}

> This policy specifies that all workloads in the mesh will only accept encrypted requests using TLS. As you can see, this authentication policy contains no `selector` specification (as it is intended to apply to all services in the mesh).

Author: done

> This policy directly instructs Pilot to program sidecar proxy to accept only mutual TLS inbound traffic. With `auto mTLS` feature enabled (see more [here](/docs/tasks/security/auto-mtls)), it will also try to configure client to use mTLS when possible. Also note, the authentication policy doesn't have effect on workload that doesn't have sidecar (e.g `httpbin.legacy`). However, `auto mTLS` also takes that into account and does not apply mTLS to the outbound traffic to the legacy service.

Reviewer: ```suggestion
- It directly instructs Istio to configure the proxies to only accept mutual TLS inbound traffic. 
- If the [auto mTLS](/docs/tasks/security/auto-mtls) feature is enabled, it also tries to configure the client to use mTLS when possible. 

The authentication policy doesn't apply to workloads without a proxy, for example `httpbin.legacy`. However, the auto mTLS feature takes that into account, and does not apply mTLS to the outbound traffic to the legacy service.
```

Author: done

> If you run the test command above, you will see all requests still succeed, except for those from client that doesn't have sidecar (`sleep.legacy`) to server with sidecar (`httpbin.foo` or `httpbin.bar`). This is expected, as the mTLS is now strictly required, but the workload without sidecar cannot comply.

Reviewer: What test command? Please repeat the command if other things are in between.
```suggestion
Run the test command again:
`<insert test command>`
You see requests still succeed, except for those from the client that doesn't have proxy, `sleep.legacy`, to the server with a proxy, `httpbin.foo` or `httpbin.bar`. This is expected because mTLS is now strictly required, but the workload without sidecar cannot comply.
```

Author: done

{{< text bash >}}
$ for from in "foo" "bar" "legacy"; do for to in "foo" "bar" "legacy"; do kubectl exec $(kubectl get pod -l app=sleep -n ${from} -o jsonpath={.items..metadata.name}) -c sleep -n ${from} -- curl "http://httpbin.${to}:8000/ip" -s -o /dev/null -w "sleep.${from} to httpbin.${to}: %{http_code}\n"; done; done
sleep.foo to httpbin.foo: 200
sleep.foo to httpbin.bar: 200
sleep.foo to httpbin.legacy: 200
sleep.bar to httpbin.foo: 200
sleep.bar to httpbin.bar: 200
sleep.bar to httpbin.legacy: 200
sleep.legacy to httpbin.foo: 000
command terminated with exit code 56
sleep.legacy to httpbin.bar: 000
command terminated with exit code 56
sleep.legacy to httpbin.legacy: 200
{{< /text >}}

> Using `httpbin` functionality, you can also verify that traffic from `sleep.foo` to `httpbin.foo` is indeed in mTLS:

Reviewer: ```suggestion
You can use `httpbin` to verify that traffic from `sleep.foo` to `httpbin.foo` is indeed in mTLS:
```

{{< text bash >}}
$ kubectl exec $(kubectl get pod -l app=sleep -n foo -o jsonpath={.items..metadata.name}) -c sleep -n foo -- curl http://httpbin.foo:8000/headers -s | grep X-Forwarded-Client-Cert
"X-Forwarded-Client-Cert": "By=spiffe://cluster.local/ns/foo/sa/httpbin;Hash=<redacted>"
{{< /text >}}

### Cleanup part 1

Remove global authentication policy and destination rules added in the session:

{{< text bash >}}
$ kubectl delete peerauthentication -n istio-system default
{{< /text >}}

## Enable mutual TLS per namespace or workload

### Namespace-wide policy

> If you only want to change mTLS for all workloads within a particular namespace, you can use a namespace-wide policy. The spec of the policy is the same as the mesh-wide policy, except it should be submitted to the namespace on whicy you want to applied. For example, to enable mTLS strict for namespace `foo`, you can do:

Reviewer: ```suggestion
To change mTLS for all workloads within a particular namespace, use a namespace-wide policy. The specification of the policy is the same as for a mesh-wide policy, but you specify the namespace it applies to under `metadata`. For example, the following peer authentication policy enables strict mTLS  for the `foo` namespace:
```

sleep.legacy to httpbin.legacy: 200
{{< /text >}}

### Enable mutual TLS per workload

> You can also set peer authentication policy for a specific workload by setting the `selector` to match to the desired workload (by labels). However, Istio cannot aggregate workload-level policies when programming the outbound mTLS to a service, you will also need to provide a DestinationRule to specify that.

Reviewer: ```suggestion
To set a peer authentication policy for a specific workload, you must configure the `selector` section and specify the labels that match the desired workload. However, Istio cannot aggregate workload-level policies for outbound mTLS traffic to a service. Configure a destination rule to manage that behavior.
```

Author: done

> For example, to set mTLS strict for `httpbin.bar`, you will need to set both peer authentication policy as well as destination rule like below:

Reviewer: ```suggestion
For example, the following peer authentication policy and destination rule enable strict mTLS for the `httpbin.bar` workload:
```

Author: done

sleep.legacy to httpbin.legacy: 200
{{< /text >}}

### Enable mutual TLS per workload

> To set a peer authentication policy for a specific workload, you must configure the `selector` section and specify the labels that match the desired workload. However, Istio cannot aggregate workload-level policies for outbound mTLS traffic to a service. Configure a destination rule to manage that behavior.

Reviewer 2: s/the desired/that
i feel a bit of weird of using "desired"

command terminated with exit code 56
{{< /text >}}

> You can further refine the mTLS settings per port. Our demo `httpbin` doesn't have multiple ports. But imagine that it does, then the following policy requires mTLS on all ports, except port `80` .

Reviewer: ```suggestion
To refine the mTLS settings per port, you must configure the `portLevelMtls` section. For example, the following peer authentication policy requires mTLS on all ports, except port `80`:
```

Author: done

{{< text bash >}}
$ cat <<EOF | kubectl apply -n bar -f -
apiVersion: "security.istio.io/v1beta1"
kind: "PeerAuthentication"
metadata:
  name: "httpbin"
  namespace: "bar"
spec:
  selector:
    matchLabels:
      app: httpbin
  mtls:
    mode: STRICT
  portLevelMtls:
    80:
      mode: DISABLE
EOF
{{< /text >}}

> As before, it should come with a destination rule:

Reviewer: ```suggestion
As before, you also need a destination rule, for example:
```

Author: done

command terminated with exit code 56
{{< /text >}}

To refine the mTLS settings per port, you must configure the `portLevelMtls` section. For example, the following peer authentication policy requires mTLS on all ports, except port `80`:

{{< text bash >}}
$ cat <<EOF | kubectl apply -n bar -f -
apiVersion: "security.istio.io/v1beta1"
kind: "PeerAuthentication"
metadata:
  name: "httpbin"
  namespace: "bar"
spec:
  selector:
    matchLabels:
      app: httpbin
  mtls:
    mode: STRICT
  portLevelMtls:
    80:
      mode: DISABLE
EOF
{{< /text >}}

> As before, you also need a destination rule, for example:

Reviewer 2: i think `for example` here is a bit redundant. you need to have DR exactly as below. IIUC

  host: httpbin.bar.svc.cluster.local
  trafficPolicy:
    tls:
      mode: ISTIO_MUTUAL
    portLevelSettings:
    - port:
        number: 8000
      tls:
        mode: DISABLE
EOF
{{< /text >}}

> {{< tip >}}

Reviewer: ```suggestion
```

Author: done

> - Port value in peer authentication policy is the *container port*, whereas in destination rule, it is the *service port*

Reviewer: ```suggestion
The port value in the peer authentication policy is the container's port. The value the destination rule is the workload's port.
```

Author: done

> - Only port that is binded to some service can be used in port-level Mtls. It will be ignored otherwise.

Reviewer: ```suggestion
You can only use `portLevelMtls` if the port is bound to a service. Istio ignores it otherwise.
```

Author: done

{{< /tip >}}

### Policy precedence

> To illustrate how a workload-specific policy takes precedence over namespace-wide policy, you can add a policy to disable mutual TLS for `httpbin.foo` as below.

Reviewer: ```suggestion
A workload-specific peer authentication policy takes precedence over a namespace-wide policy. You can test this behavior if you add a policy to disable mutual TLS for the `httpbin.foo` workload, for example:
```

Author: done

Remove policies and destination rules created in the above steps:

{{< text bash >}}

> $ kubectl delete peerauthentication default overwrite-example -n foo

Reviewer 2: we might consider add a short name later. out of scope

200
{{< /text >}}

> Now, add a policy that requires end-user JWT for `ingressgateway`.

Reviewer: ```suggestion
Now, add a request authentication policy that requires end-user JWT for the ingress gateway.
```

Author: done

{{< text bash >}}
kubectl apply -f - <<EOF
apiVersion: "security.istio.io/v1beta1"
kind: "RequestAuthentication"
metadata:
  name: "jwt-example"
  namespace: istio-system
spec:
  selector:
    matchLabels:
      istio: ingressgateway
  jwtRules:
  - issuer: "testing@secure.istio.io"
    jwksUri: "{{< github_file >}}/security/tools/jwt/samples/jwks.json"
EOF
{{< /text >}}

> {{< tip >}}

Reviewer: ```suggestion
```

Author: done

> - The policy need to be submitted in the same namespace as the workload it selects (`ingressgateway`). In this case, it is `istio-system`.

Reviewer: ```suggestion
Submit the policy in the same namespace as the workload it selects, `ingressgateway` in this case. The namespace you need to specify is then `istio-system`.
```

Author: done

> {{< /tip >}}

Reviewer: ```suggestion
```

Author: done

> The policy above means if a token is provided in the Authorization header (implicitly default location), then it will be validated using the public key set at "{{< github_file >}}/security/tools/jwt/samples/jwks.json". Request will be rejected if the bearing token is invalid. Howerver, request without token is still accepted. To verify this behavior, retry the request without token, with bad token and with a valid token as followed:

Reviewer: ```suggestion
If you provide a token in the authorization header, its implicitly default location, Istio validates the token using the [public key set]({{< github_file >}}/security/tools/jwt/samples/jwks.json), and rejects requests if the bearing token is invalid. However, requests without tokens are accepted. To observe this behavior, retry the request without a token, with a bad token, and with a valid token:
```

Author: done

200
{{< /text >}}

Now, add a request authentication policy that requires end-user JWT for the ingress gateway.

{{< text bash >}}
$ kubectl apply -f - <<EOF
apiVersion: "security.istio.io/v1beta1"
kind: "RequestAuthentication"
metadata:
  name: "jwt-example"
  namespace: istio-system
spec:
  selector:
    matchLabels:
      istio: ingressgateway
  jwtRules:
  - issuer: "testing@secure.istio.io"
    jwksUri: "{{< github_file >}}/security/tools/jwt/samples/jwks.json"
EOF
{{< /text >}}

> Submit the policy in the same namespace as the workload it selects, `ingressgateway` in this case. The namespace you need to specify is then `istio-system`.

Reviewer 2: Submit policy seems a non conventional usage, compared to the rest of the wording.

Reviewer: I agree, WDYT about:
```suggestion
Apply the policy to the namespace of the workload it selects, `ingressgateway` in this case. The namespace you need to specify is then `istio-system`.
```

> Submit the policy in the same namespace as the workload it selects, `ingressgateway` in this case. The namespace you need to specify is then `istio-system`.

Reviewer 2: this sentence seems by itself is a bit off, disconnected with context.

> 

Reviewer 2: why two newlines

> If you provide a token in the authorization header, its implicitly default location, Istio validates the token using the [public key set]({{< github_file >}}/security/tools/jwt/samples/jwks.json), and rejects requests if the bearing token is invalid. However, requests without tokens are accepted. To observe this behavior, retry the request without a token, with a bad token, and with a valid token:

Reviewer 2: `its impolicit default location` this `its` is vague, default location of what?

> If you provide a token in the authorization header, its implicitly default location, Istio validates the token using the [public key set]({{< github_file >}}/security/tools/jwt/samples/jwks.json), and rejects requests if the bearing token is invalid. However, requests without tokens are accepted. To observe this behavior, retry the request without a token, with a bad token, and with a valid token:

Reviewer 2: bearing token or bearer token? i have only seen latter as common usage.

and introduce a new term in this paragraph again. either replace `a token` -> `a bearer token.` all the places in this para.

Reviewer: Great catch, bearer token is jargon and I missed it, so it's likely your readers will too. Try:
```suggestion
If you provide a token in the authorization header, its implicitly default location, Istio validates the token using the [public key set]({{< github_file >}}/security/tools/jwt/samples/jwks.json), and rejects requests if the bearer token is invalid. However, requests without tokens are accepted. To observe this behavior, retry the request without a token, with a bad token, and with a valid token:
```
We can add a glossary entry later for the term.

You can also add a JWT policy to an ingress gateway (e.g., service `istio-ingressgateway.istio-system.svc.cluster.local`).
This is often used to define a JWT policy for all services bound to the gateway, instead of for individual services.

### Enforce valid token must be presented.

To reject request without valid token, you will need to add an authorization rule to `deny` request without `requestPrincipal` as below. Request principal is available only when a valid JWT token is provided, so the rule is equivalent to deny requests without valid token.

{{< text bash >}}
kubectl apply -f - <<EOF
apiVersion: "security.istio.io/v1beta1"
kind: "AuthorizationPolicy"
metadata:
  name: "frontend-ingress"
  namespace: istio-system
  labels:

>     demo.istio.io: security

Reviewer 3: why is this label?

Author: Want to make the clean up command simpler. Remove it for now

spec:
  selector:
    matchLabels:
      istio: ingressgateway
  action: DENY
  rules:
  - from:
    - source:
        notRequestPrincipals: ["*"]
EOF
{{< /text >}}

Retry request without token, you should see it now fail with error code 403.

{{< text bash >}}
$ curl $INGRESS_HOST/headers -s -o /dev/null -w "%{http_code}\n"
403
{{< /text >}}

### Enforce valid token per-path

You can also refine the requirement per host, path, method etc with authorization. For example, you can change the authorization policy in the last section to only require JWT on `/headers`. After applying this authorization rule, request to `$INGRESS_HOST/headers` will fail with 403 code. But for all other path, e.g `$INGRESS_HOST/ip`, it should succeed. Please see other [tasks with authorization]() to learn more about other ways to refine access control with authoriation policy.

{{< text bash >}}
kubectl apply -f - <<EOF
apiVersion: "security.istio.io/v1beta1"
kind: "AuthorizationPolicy"
metadata:

>   name: "frontend-ingress"

Reviewer 3: could you also add curl commands to verify the authZ policy on path /headers and another path (e.g. /ip)?

Author: done

  namespace: istio-system
  labels:

>     demo.istio.io: security

Reviewer 3: same here.

spec:
  selector:
    matchLabels:
      istio: ingressgateway
  action: DENY
  rules:
  - from:
    - source:
        notRequestPrincipals: ["*"]
    to:
    - operation:
        paths: ["/headers"]
EOF
{{< /text >}}

### Cleanup part 3

1. Remove authentication policy:

    {{< text bash >}}

>     $ kubectl -n istio-system delete requestauthentication jwt-example

Reviewer 3: Need to also remove the authorization policy.

Author: done

You can also add a JWT policy to an ingress gateway (e.g., service `istio-ingressgateway.istio-system.svc.cluster.local`).
This is often used to define a JWT policy for all services bound to the gateway, instead of for individual services.

> ### Enforce valid token must be presented.

Reviewer: ```suggestion
### Require a valid token
```

Author: done

> To reject request without valid token, you will need to add an authorization rule to `deny` request without `requestPrincipal` as below. Request principal is available only when a valid JWT token is provided, so the rule is equivalent to deny requests without valid token.

Reviewer: ```suggestion
To reject requests without valid tokens, add an authorization rule specifying a `DENY` action for requests without request principals, shown as `notRequestPrincipals: ["*"]` in the following example. Request principals are available only when valid JWT tokens are provided. The rule therefore denies requests without valid tokens.
```

Author: done

{{< text bash >}}
kubectl apply -f - <<EOF
apiVersion: "security.istio.io/v1beta1"
kind: "AuthorizationPolicy"
metadata:
  name: "frontend-ingress"
  namespace: istio-system
  labels:
    demo.istio.io: security
spec:
  selector:
    matchLabels:
      istio: ingressgateway
  action: DENY
  rules:
  - from:
    - source:
        notRequestPrincipals: ["*"]
EOF
{{< /text >}}

> Retry request without token, you should see it now fail with error code 403.

Reviewer: ```suggestion
Retry the request without a token. The request now fails with error code `403`:
```

Author: done

{{< text bash >}}
$ curl $INGRESS_HOST/headers -s -o /dev/null -w "%{http_code}\n"
403
{{< /text >}}

> ### Enforce valid token per-path

Reviewer: ```suggestion
### Require valid tokens per-path
```

Author: done

> You can also refine the requirement per host, path, method etc with authorization. For example, you can change the authorization policy in the last section to only require JWT on `/headers`. After applying this authorization rule, request to `$INGRESS_HOST/headers` will fail with 403 code. But for all other path, e.g `$INGRESS_HOST/ip`, it should succeed. Please see other [tasks with authorization]() to learn more about other ways to refine access control with authoriation policy.

Reviewer: I removed the link to the other authorization tasks since it was empty.

```suggestion
To refine authorization with a token requirement per host, path, or method, change the authorization policy to only require JWT on `/headers`. When this authorization rule takes effect, requests to `$INGRESS_HOST/headers` fail with the error code `403`. Requests to all other paths succeed, for example `$INGRESS_HOST/ip`. 
```

Author: done

You can also add a JWT policy to an ingress gateway (e.g., service `istio-ingressgateway.istio-system.svc.cluster.local`).
This is often used to define a JWT policy for all services bound to the gateway, instead of for individual services.

### Require a valid token

> To reject requests without valid tokens, add an authorization rule specifying a `DENY` action for requests without request principals, shown as `notRequestPrincipals: ["*"]` in the following example. Request principals are available only when valid JWT tokens are provided. The rule therefore denies requests without valid tokens.

Reviewer 2: s/authorizatoin rule/authorization policy rule

Reviewer: ```suggestion
To reject requests without valid tokens, add an authorization policy with a rule specifying a `DENY` action for requests without request principals, shown as `notRequestPrincipals: ["*"]` in the following example. Request principals are available only when valid JWT tokens are provided. The rule therefore denies requests without valid tokens.
```

> To reject requests without valid tokens, add an authorization rule specifying a `DENY` action for requests without request principals, shown as `notRequestPrincipals: ["*"]` in the following example. Request principals are available only when valid JWT tokens are provided. The rule therefore denies requests without valid tokens.

Reviewer 2: in the following example, can be put in the end of paragraph. here is jump-in and then switch to explaining.

Reviewer: You are correct, but I thought that was a minor issue. I'm fine if you move it to the end:
```suggestion
To reject requests without valid tokens, add an authorization rule specifying a `DENY` action for requests without request principals, shown as `notRequestPrincipals: ["*"]`. Request principals are available only when valid JWT tokens are provided. The rule therefore denies requests without valid tokens as seen in the following example:
```
